[av][Android] Bump `com.google.android.exoplayer:exoplayer` from `2.9.2` to `2.13.3`

bbarthec opened the pull request:
# Why

Resolves [ENG-2858](https://linear.app/expo/issue/ENG-2858/remove-jcenter)
Continues https://github.com/expo/expo/pull/16122

# How

- Removed `https://github.com/yqritc/Android-ScalableVideoView` by copying the parts into the `expo-av` codebase as it's only available on `jcenter()`
- Bumped the `exoplayer` to the smallest version available on google maven.
- Tweaked the codebase to work with the new version of exoplayer.

# Test Plan

I've succeeded in compiling `bare-expo` and running the `ncl`#`video` screen with both remote and local videos playing nicely.

# TODO

- [x] Ensure `Expo Go` works
- [ ] Backport changes if necessary

Kudo commented on android/expoview/build.gradle:
```diff
@@ -99,6 +98,7 @@ repositories {
 
 android {
   compileSdkVersion safeExtGet("compileSdkVersion", 30)
+  ndkVersion "23.0.7421159"
```
any reason we should upgrade ndk version here?

nomi9995 replied:
yes, no need this line, i have checked it on my local

Kudo reviewed (approved):
LGTM. i think android client test error is from expo go, maybe you will update in this pr or another pr.

mnightingale commented:
I've not checked but looking at the release notes for ExoPlayer this possibly resolves a few open av issues (#16139 and maybe #16439).

Is there any reason not to go for a more recent ExoPlayer version (latest is currently 2.17.0, about 10 months more recent than 2.13.3), this PR already makes pretty much all of the required API changes from the last few years so it's not much more work.

A few changes would be required:

- buildMediaSource handle TYPE_RTSP
- A few renamed classes or constants
- Likely upgrade okhttp + okhttp-urlconnection so it uses the same version as ExoPlayer
- A few changes to SharedCookiesDataSourceFactory and CustomHeadersOkHttpDataSourceFactory - might also be able to simplify those using okhttp interceptor, else I think apis are exposed so setDefaultRequestProperties can be used

There are also usages of a few deprecated methods that could be cleaned up.

mnightingale commented on packages/expo-av/android/build.gradle:
```diff
@@ -74,11 +74,9 @@ dependencies {
 
   // Newer version introduces dependency versions conflict
   // on 'com.android.support:support-annotations'
```
```suggestion
```

It appears there isn't a conflict anymore, hence this PR is upgrading exoplayer.

bbarthec committed:
[av][Expo Go][Android] Bump `com.google.android.exoplayer.*` to version `2.13.3`

- Bump `com.google.android.exoplayer:exoplayer` and `com.google.android.exoplayer:exoplayer-okhttp` from `2.9.2` to `2.13.3`
- Remove unused `com.android.support:support-annotations` dependecy
- Remove unused `supportLibVersion` variable from gradle
- Remove `jcenter` repository from `expoview` project

bbarthec committed:
[av][Android] Copy-paste files from `com.yqritc:android-scalablevideoview`

- `com.yqritc:android-scalablevideoview` is only available on `jcenter` repository and we're removing it, so I've copy-pasted classes that we rely on in `expo-av`

bbarthec committed:
[av][Android] Propagate `com.yqritc:android-scalablevideoview` lib replacement

bbarthec committed:
[av][Android] Adapt the ExoPlayer code to the newer version

- Replace `ExtractorMediaSource.EventListener` with `MediaSourceEventListener`
- Replace `SimpleExoPlayer.VideoListener` with `VideoListener`
- Replace `AdaptiveMediaSourceEventListener` with `MediaSourceEventListener`
- Refactor `mSimpleExoPlayer` construction
- Adapt creating `MediaSource`
- Cleanup the codebase

bbarthec committed:
[no ci] Upgrade CHANGELOG

bbarthec committed:
[apps] Remove `jcenter` from android apps

Remove `jcenter` from `bare-expo` and `bare-sandbox` apps

bbarthec committed:
[Expo Go][Android] Remove `jcenter`

bbarthec committed:
[expo-av][Android] Backport changes to SDK 42,43,44

bbarthec committed:
Update yarn.lock to make CI pass

dantxal commented:
Hello bbarthec!! 

I've been trying to test your solution, I tried to build by linking the dependencies of my repo to your branch on github, but that didn't work.

Now I see that you merged it, could you give me the instructions to get it running on my repo?

This is the repo I've been trying to test it on: [dantxal/ExpoAvIssue](https://github.com/dantxal/ExpoAvIssue).

This is the issue I talked about, for reference: #16323 

Thanks for the good work you've been doing on expo!

bbarthec commented:
@dantxal, this problem you're reporting is very generic and I have no way to help you with it 😞 
I'm going to publish this change to the npm with the `next` tag and you'll be able to test it that way 👀

bbarthec commented:
I've just published [`expo-av@11.0.0`](https://www.npmjs.com/package/expo-av/v/11.0.0?activeTab=versions) with the `next` tag on `npm` containing this PR and many more (see [CHANGELOG](https://github.com/expo/expo/blob/main/packages/expo-av/CHANGELOG.md#1100--2022-03-03)) 🎉 
Please try it and feel free to provide any kind of feedback you want (including new issues, because something is broken out of the blue) 😁

dantxal commented:
@bbarthec You're right, it didn't solve my problem.

To anyone else trying to run `expo-av@11.0.0`: 
You might run into this error on expo package: 
```sh
Module was compiled with an incompatible version of Kotlin.
```

All I had to do to get it running was:

1. Navigate into expo package within node_modules (relative path: `node_modules/expo/android/build.gradle`);
2. Change line 24 from 
```java
    classpath("org.jetbrains.kotlin:kotlin-gradle-plugin:${safeExtGet('kotlinVersion', '1.4.1')}")
```
into
```java
    classpath("org.jetbrains.kotlin:kotlin-gradle-plugin:${safeExtGet('kotlinVersion', '1.6.0')}")
```
3. Run `yarn android` again. ( I ran `cd android && ./gradlew clean && cd ..` just to be safe) .

eeshankeni commented:
Please consider implementing https://github.com/expo/expo/pull/16455 to expo-av 11 as well

hirbod commented:
So far, expo-av 11 is working fine for me on Android. What still is so extremely annoying is the massive pauseAsync delay on iOS and the framedrops caused by it (from 60 to 20). Android does not have this issues.

KrisLau commented:
@bbarthec @Kudo Thanks for all the awesome work! I've been working on a fatal error I've been getting on my project and realized that the `exoplayer` version being too low is causing a fatal error when used with `react-native-track-player` which uses v2.18.1. Are there any plans to upgrade to that version any time soon or ado you guys have a workaround for that?

Issue with repro: https://github.com/expo/expo/issues/18937
